Optimize FindFirstCharToEncode for JavaScriptEncoder.Default using Ssse3 intrinsics

gfoidl opened the pull request:
## Description

Follow up to https://github.com/dotnet/corefx/pull/41933

* For the vectorized char-path 2\*8 chars are combined to one byte-vector (16 elements), then the byte-path for finding the first element, that needs to be escaped is taken. If there are less than 2*8 chars, a dummy char-vector is used, that is known to be be valid, i.e. it must not be escaped (here `A` is chosen)
(cf. https://github.com/dotnet/corefx/pull/41845#discussion_r336728111)

* The "escaping mask" is created via bit-masks. For an outline of the algorithm see https://github.com/dotnet/corefx/pull/41845#discussion_r336745516 (and / or comments in the code).
For the shuffling SSSE3 is needed (for a comment on SSSE3 see also https://github.com/dotnet/corefx/pull/41845#discussion_r337580425).

These optimization are applied to `JavaScriptEncoder.Default` (= `DefaultJavaScriptEncoderBasicLatin`), as suggested by @ahsonkhan in https://github.com/dotnet/corefx/pull/41845#discussion_r337903046.

For the other two mentioned encoders, the bit-mask won't work (in a performant way), as some values above `0x7F` are valid. So to handle them properly, lots of overhead have to be introduced, that's why I left them out (in this PR, maybe later when I can think more about it).

## Benchmarks

Benchmarks are based on the ones from https://github.com/dotnet/corefx/pull/41845

<details>
    <summary>Machine info</summary>

A linux-vm was used.
```
BenchmarkDotNet=v0.11.5, OS=ubuntu 18.04
Intel Xeon CPU 2.00GHz, 1 CPU, 2 logical cores and 1 physical core
```

</details>

<details>
   <summary>Test_EscapingBenchmark</summary>

```
|   Method |        TestStringData |      Mean |     Error |    StdDev | Ratio | RatioSD |
|--------- |---------------------- |----------:|----------:|----------:|------:|--------:|
|   Master |            (1, -1, r) |  5.660 ns | 0.1006 ns | 0.0941 ns |  1.00 |    0.00 |
| PR_42073 |            (1, -1, r) |  5.318 ns | 0.0420 ns | 0.0393 ns |  0.94 |    0.02 |
|          |                       |           |           |           |       |         |
|   Master |          (1, 0, &lt;) |  5.072 ns | 0.0081 ns | 0.0071 ns |  1.00 |    0.00 |
| PR_42073 |          (1, 0, &lt;) |  5.113 ns | 0.0132 ns | 0.0123 ns |  1.01 |    0.00 |
|          |                       |           |           |           |       |         |
|   Master | (100,(...)kuwu) [111] | 48.072 ns | 0.4432 ns | 0.3929 ns |  1.00 |    0.00 |
| PR_42073 | (100,(...)kuwu) [111] | 24.007 ns | 0.0502 ns | 0.0419 ns |  0.50 |    0.00 |
|          |                       |           |           |           |       |         |
|   Master |  (15, (...)dabu) [25] | 21.191 ns | 0.0343 ns | 0.0268 ns |  1.00 |    0.00 |
| PR_42073 |  (15, (...)dabu) [25] | 13.933 ns | 0.0442 ns | 0.0391 ns |  0.66 |    0.00 |
|          |                       |           |           |           |       |         |
|   Master |  (16, (...)sqfa) [26] | 14.327 ns | 0.0210 ns | 0.0186 ns |  1.00 |    0.00 |
| PR_42073 |  (16, (...)sqfa) [26] |  8.877 ns | 0.0458 ns | 0.0428 ns |  0.62 |    0.00 |
|          |                       |           |           |           |       |         |
|   Master |  (17, (...)aabr) [27] | 16.068 ns | 0.0447 ns | 0.0373 ns |  1.00 |    0.00 |
| PR_42073 |  (17, (...)aabr) [27] | 10.433 ns | 0.0337 ns | 0.0299 ns |  0.65 |    0.00 |
|          |                       |           |           |           |       |         |
|   Master |           (2, -1, dd) |  6.631 ns | 0.0198 ns | 0.0175 ns |  1.00 |    0.00 |
| PR_42073 |           (2, -1, dd) |  6.622 ns | 0.0323 ns | 0.0302 ns |  1.00 |    0.01 |
|          |                       |           |           |           |       |         |
|   Master |  (32, (...)dfpn) [42] | 20.296 ns | 0.2466 ns | 0.2186 ns |  1.00 |    0.00 |
| PR_42073 |  (32, (...)dfpn) [42] | 10.287 ns | 0.2252 ns | 0.2107 ns |  0.51 |    0.01 |
|          |                       |           |           |           |       |         |
|   Master |         (4, -1, negs) |  9.686 ns | 0.0922 ns | 0.0863 ns |  1.00 |    0.00 |
| PR_42073 |         (4, -1, negs) |  9.128 ns | 0.0523 ns | 0.0463 ns |  0.94 |    0.01 |
|          |                       |           |           |           |       |         |
|   Master |      (7, -1, netggni) | 14.083 ns | 0.0461 ns | 0.0385 ns |  1.00 |    0.00 |
| PR_42073 |      (7, -1, netggni) | 14.761 ns | 0.0426 ns | 0.0356 ns |  1.05 |    0.00 |
|          |                       |           |           |           |       |         |
|   Master |    (7, 0, &lt;etggni) |  5.290 ns | 0.0194 ns | 0.0172 ns |  1.00 |    0.00 |
| PR_42073 |    (7, 0, &lt;etggni) |  5.101 ns | 0.0280 ns | 0.0262 ns |  0.96 |    0.01 |
|          |                       |           |           |           |       |         |
|   Master |    (7, 1, n&lt;tggni) |  6.439 ns | 0.0443 ns | 0.0414 ns |  1.00 |    0.00 |
| PR_42073 |    (7, 1, n&lt;tggni) |  6.427 ns | 0.0451 ns | 0.0400 ns |  1.00 |    0.01 |
|          |                       |           |           |           |       |         |
|   Master |    (7, 2, ne&lt;ggni) |  8.023 ns | 0.0340 ns | 0.0301 ns |  1.00 |    0.00 |
| PR_42073 |    (7, 2, ne&lt;ggni) |  7.958 ns | 0.1538 ns | 0.1363 ns |  0.99 |    0.02 |
|          |                       |           |           |           |       |         |
|   Master |    (7, 3, net&lt;gni) |  9.177 ns | 0.0694 ns | 0.0615 ns |  1.00 |    0.00 |
| PR_42073 |    (7, 3, net&lt;gni) |  8.808 ns | 0.0230 ns | 0.0192 ns |  0.96 |    0.01 |
|          |                       |           |           |           |       |         |
|   Master |    (7, 4, netg&lt;ni) | 10.564 ns | 0.0143 ns | 0.0120 ns |  1.00 |    0.00 |
| PR_42073 |    (7, 4, netg&lt;ni) | 10.590 ns | 0.0143 ns | 0.0127 ns |  1.00 |    0.00 |
|          |                       |           |           |           |       |         |
|   Master |    (7, 5, netgg&lt;i) | 12.577 ns | 0.0185 ns | 0.0154 ns |  1.00 |    0.00 |
| PR_42073 |    (7, 5, netgg&lt;i) | 11.106 ns | 0.0306 ns | 0.0271 ns |  0.88 |    0.00 |
|          |                       |           |           |           |       |         |
|   Master |    (7, 6, netggn&lt;) | 14.034 ns | 0.0385 ns | 0.0341 ns |  1.00 |    0.00 |
| PR_42073 |    (7, 6, netggn&lt;) | 12.251 ns | 0.0163 ns | 0.0145 ns |  0.87 |    0.00 |
|          |                       |           |           |           |       |         |
|   Master |     (8, -1, jgnavpkd) | 10.966 ns | 0.0180 ns | 0.0150 ns |  1.00 |    0.00 |
| PR_42073 |     (8, -1, jgnavpkd) |  9.062 ns | 0.0167 ns | 0.0148 ns |  0.83 |    0.00 |
|          |                       |           |           |           |       |         |
|   Master |   (8, 0, &lt;gnavpkd) | 11.446 ns | 0.0321 ns | 0.0301 ns |  1.00 |    0.00 |
| PR_42073 |   (8, 0, &lt;gnavpkd) |  8.655 ns | 0.0157 ns | 0.0140 ns |  0.76 |    0.00 |
|          |                       |           |           |           |       |         |
|   Master |   (8, 1, j&lt;navpkd) | 10.585 ns | 0.0197 ns | 0.0164 ns |  1.00 |    0.00 |
| PR_42073 |   (8, 1, j&lt;navpkd) |  8.674 ns | 0.0202 ns | 0.0189 ns |  0.82 |    0.00 |
|          |                       |           |           |           |       |         |
|   Master |   (8, 2, jg&lt;avpkd) | 10.568 ns | 0.0133 ns | 0.0118 ns |  1.00 |    0.00 |
| PR_42073 |   (8, 2, jg&lt;avpkd) |  8.663 ns | 0.0235 ns | 0.0220 ns |  0.82 |    0.00 |
|          |                       |           |           |           |       |         |
|   Master |   (8, 3, jgn&lt;vpkd) | 10.583 ns | 0.0154 ns | 0.0144 ns |  1.00 |    0.00 |
| PR_42073 |   (8, 3, jgn&lt;vpkd) |  8.654 ns | 0.0115 ns | 0.0102 ns |  0.82 |    0.00 |
|          |                       |           |           |           |       |         |
|   Master |   (8, 4, jgna&lt;pkd) | 10.986 ns | 0.0200 ns | 0.0187 ns |  1.00 |    0.00 |
| PR_42073 |   (8, 4, jgna&lt;pkd) |  8.651 ns | 0.0104 ns | 0.0097 ns |  0.79 |    0.00 |
|          |                       |           |           |           |       |         |
|   Master |   (8, 5, jgnav&lt;kd) | 10.579 ns | 0.0145 ns | 0.0121 ns |  1.00 |    0.00 |
| PR_42073 |   (8, 5, jgnav&lt;kd) |  8.652 ns | 0.0099 ns | 0.0093 ns |  0.82 |    0.00 |
|          |                       |           |           |           |       |         |
|   Master |   (8, 6, jgnavp&lt;d) | 10.986 ns | 0.0350 ns | 0.0311 ns |  1.00 |    0.00 |
| PR_42073 |   (8, 6, jgnavp&lt;d) |  8.656 ns | 0.0127 ns | 0.0112 ns |  0.79 |    0.00 |
|          |                       |           |           |           |       |         |
|   Master |   (8, 7, jgnavpk&lt;) | 10.963 ns | 0.0139 ns | 0.0116 ns |  1.00 |    0.00 |
| PR_42073 |   (8, 7, jgnavpk&lt;) |  8.660 ns | 0.0096 ns | 0.0085 ns |  0.79 |    0.00 |
|          |                       |           |           |           |       |         |
|   Master |    (9, -1, csvobsdxs) | 11.934 ns | 0.0233 ns | 0.0218 ns |  1.00 |    0.00 |
| PR_42073 |    (9, -1, csvobsdxs) | 10.200 ns | 0.0113 ns | 0.0100 ns |  0.86 |    0.00 |
```

</details>

<details>
    <summary>Perf.Basic</summary>

#### Master

```
|          Method | Formatted | SkipValidation | DataSize |     Mean |     Error |    StdDev |   Median |      Min |      Max |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------- |---------- |--------------- |--------- |---------:|----------:|----------:|---------:|---------:|---------:|-------:|------:|------:|----------:|
| WriteBasicUtf16 |     False |           True |       10 | 1.391 us | 0.0156 us | 0.0146 us | 1.388 us | 1.367 us | 1.419 us | 0.0056 |     - |     - |     120 B |
```

#### PR_42073 (this one)

```
|          Method | Formatted | SkipValidation | DataSize |     Mean |     Error |    StdDev |   Median |      Min |      Max |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------------- |---------- |--------------- |--------- |---------:|----------:|----------:|---------:|---------:|---------:|-------:|------:|------:|----------:|
| WriteBasicUtf16 |     False |           True |       10 | 1.316 us | 0.0034 us | 0.0026 us | 1.315 us | 1.311 us | 1.320 us | 0.0052 |     - |     - |     120 B |

```

</details>

<details>
    <summary>EscapingWriter</summary>

#### Master

```
|               Method | DataLength | NegativeIndex |      Mean |    Error |   StdDev |    Median |       Min |       Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------- |----------- |-------------- |----------:|---------:|---------:|----------:|----------:|----------:|------:|------:|------:|----------:|
| NeedsEscapingCurrent |         32 |            -1 |  89.13 us | 1.886 us | 1.764 us |  88.46 us |  87.40 us |  92.88 us |     - |     - |     - |         - |
| NeedsEscapingCurrent |         32 |            12 | 208.18 us | 0.407 us | 0.340 us | 208.17 us | 207.67 us | 208.94 us |     - |     - |     - |         - |
```

#### PR_42073 (this one)

```
|               Method | DataLength | NegativeIndex |      Mean |    Error |   StdDev |    Median |       Min |       Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------- |----------- |-------------- |----------:|---------:|---------:|----------:|----------:|----------:|------:|------:|------:|----------:|
| NeedsEscapingCurrent |         32 |            -1 |  74.71 us | 0.337 us | 0.263 us |  74.66 us |  74.29 us |  75.16 us |     - |     - |     - |         - |
| NeedsEscapingCurrent |         32 |            12 | 203.55 us | 1.189 us | 1.054 us | 203.18 us | 202.71 us | 206.46 us |     - |     - |     - |         - |
```

</details>

<details>
    <summary>WriteJson</summary>

#### Master

```
|                    Type |            Method |         Mean |        Error |       StdDev |       Median |          Min |          Max |   Gen 0 |   Gen 1 |   Gen 2 | Allocated |
|------------------------ |------------------ |-------------:|-------------:|-------------:|-------------:|-------------:|-------------:|--------:|--------:|--------:|----------:|
|          IndexViewModel | SerializeToString |  42,653.8 ns |    365.08 ns |    323.63 ns |  42,589.4 ns |  42,135.8 ns |  43,221.9 ns |  1.2052 |       - |       - |   25568 B |
|                Location | SerializeToString |   1,795.5 ns |     14.36 ns |     11.99 ns |   1,793.5 ns |   1,780.8 ns |   1,825.0 ns |  0.0290 |       - |       - |     584 B |
|          LoginViewModel | SerializeToString |     802.8 ns |     12.37 ns |     11.57 ns |     803.8 ns |     784.7 ns |     824.8 ns |  0.0161 |       - |       - |     344 B |
| MyEventsListerViewModel | SerializeToString | 921,083.1 ns | 11,547.56 ns | 10,801.59 ns | 918,623.4 ns | 898,673.1 ns | 941,143.2 ns | 44.1176 | 44.1176 | 44.1176 |  324336 B |
```

#### PR_42073 (this one)

```
|                    Type |            Method |         Mean |       Error |      StdDev |       Median |          Min |          Max |   Gen 0 |   Gen 1 |   Gen 2 | Allocated |
|------------------------ |------------------ |-------------:|------------:|------------:|-------------:|-------------:|-------------:|--------:|--------:|--------:|----------:|
|          IndexViewModel | SerializeToString |  39,136.4 ns |   176.95 ns |   147.76 ns |  39,128.9 ns |  38,954.2 ns |  39,407.3 ns |  1.2500 |       - |       - |   25608 B |
|                Location | SerializeToString |   1,787.5 ns |    11.78 ns |    10.44 ns |   1,784.1 ns |   1,775.6 ns |   1,815.9 ns |  0.0286 |       - |       - |     584 B |
|          LoginViewModel | SerializeToString |     754.7 ns |     8.18 ns |     6.83 ns |     752.6 ns |     748.7 ns |     771.6 ns |  0.0152 |       - |       - |     344 B |
| MyEventsListerViewModel | SerializeToString | 892,401.1 ns | 4,694.61 ns | 4,161.65 ns | 891,426.9 ns | 886,616.1 ns | 901,364.4 ns | 45.1389 | 45.1389 | 45.1389 |  324824 B |
```

</details>

<details>
    <summary>SerializingNuget</summary>

#### Master

```
|            Method |    Mean |    Error |   StdDev |  Median |     Min |     Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------ |--------:|---------:|---------:|--------:|--------:|--------:|------:|------:|------:|----------:|
| SerializeToString | 1.043 s | 0.0049 s | 0.0044 s | 1.043 s | 1.035 s | 1.050 s |     - |     - |     - |  787.3 MB |
```

#### PR_42073 (this one)

```
|            Method |     Mean |   Error |  StdDev |   Median |      Min |      Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------ |---------:|--------:|--------:|---------:|---------:|---------:|------:|------:|------:|----------:|
| SerializeToString | 955.2 ms | 5.17 ms | 4.32 ms | 955.1 ms | 946.8 ms | 961.5 ms |     - |     - |     - |  787.3 MB |
```

</details>

gfoidl commented:
Messed something up while copying from one vm (local build) to the repo. Will fix that build failure.

ahsonkhan commented on src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/DefaultJavaScriptEncoderBasicLatin.cs:
```diff
@@ -285,5 +378,31 @@ public override unsafe bool TryEncodeUnicodeScalar(int unicodeScalar, char* buff
         private static bool NeedsEscaping(char value) => value > LastAsciiCharacter || AllowList[value] == 0;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int NeedsEscaping(Vector128<sbyte> sourceValue)
```
> @ahsonkhan please help me figure out what causes the [build-failure]

You are seeing the failure particularly on the netfx leg. Vector128 is not available there (alone with the Sse2/Ssse3/other hardware intrinsics. Those are only available on netcoreapp.

You need to put this code under the `#if NETCOREAPP` block (similar to the other code that references those types). Or move it to a file that is conditionally included like Sse2Helper.cs:

https://github.com/dotnet/corefx/blob/bd18c12df9e5dd365a8b708cc2e7e0488b8177c1/src/System.Text.Encodings.Web/src/System.Text.Encodings.Web.csproj#L25-L27

gfoidl replied:
Ahh, thanks! So I know to check this with `./build.sh -f netfx`.
Will fix that.

gfoidl replied:
For the failed tests: when `utf8Text` is empty, we get a nullptr from `fixed` (so far nothing new to me), but then at subtracting from this nullptr, it will overflow, so the vectorized path is taken and loading the vector it crashed. Good that there are tests -- and I've learned that nullptr need more care than I thought.

ahsonkhan replied:
Yep, the current implementation in master is safe against the null ptr access because all such operations are within length checks. Hence the tests :)

ahsonkhan commented on src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs:
```diff
@@ -909,20 +911,31 @@ private byte[] GetAsciiEncoding(byte value)
         }
 
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void EnsureAsciiCacheInitialized()
```
Can you do a perf test or this, especially for small strings?

I don't know if we should explicitly mark this method as no inlining since presumably, this should be a quick check and return.

Would it be better to move the quick check into the caller, or maybe, move the initialization for-loop into a non-inlinable method, but keep this method undecorated?

gfoidl replied:
I'll investigate this.

gfoidl replied:
I think it's better to handle this in a separate PR. So I'll remove this here, to keep this PR for the SSSE3-part only.

gfoidl replied:
-> https://github.com/dotnet/corefx/pull/42177

ahsonkhan commented on src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/DefaultJavaScriptEncoderBasicLatin.cs:
```diff
@@ -285,5 +378,31 @@ public override unsafe bool TryEncodeUnicodeScalar(int unicodeScalar, char* buff
+            // Check if any of the 16 bytes need to be escaped.
+            Vector128<sbyte> mask = Ssse3.IsSupported
+                ? Ssse3Helper.CreateEscapingMask_DefaultJavaScriptEncoderBasicLatin(sourceValue)
+                : Sse2Helper.CreateEscapingMask_DefaultJavaScriptEncoderBasicLatin(sourceValue);
```
I haven't looked at the rest of the code changes yet, but don't we need an `Sse2.IsSupported` check as well?

gfoidl replied:
This method is only called from places where `Sse2.IsSupported` is already checked. So if SSSE3 is not supported, it is ensured that SSE2 is available. I'll add a `Debug.Assert` to make this more obvious.

gfoidl commented:
> Could you compare against master and this PR instead, specifically for `Test_EscapingBenchmark`?

All benchmarks are against master, except `Test_Escaping` (see below).
It's a local corefx-build (first master then this branch), and benchmarks are run with
```bash
dotnet run -c Release -f netcoreapp3.0 --corerun $corerunPath --filter System.Text.Json.Tests.Perf_Basic.WriteBasicUtf16
dotnet run -c Release -f netcoreapp3.0 --corerun $corerunPath --filter 'System.Text.Json.Serialization.Tests.WriteJson<LoginViewModel>.SerializeToString' 'System.Text.Json.Serialization.Tests.WriteJson<Location>.SerializeToString' 'System.Text.Json.Serialization.Tests.WriteJson<IndexViewModel>.SerializeToString' 'System.Text.Json.Serialization.Tests.WriteJson<MyEventsListerViewModel>.SerializeToString' --join
dotnet run -c Release -f netcoreapp3.0 --corerun $corerunPath --filter TestEscapingWriter.NeedsEscapingCurrent
dotnet run -c Release -f netcoreapp3.0 --corerun $corerunPath --filter Test_SerializingNuget.SerializeToString
```
from dotnet/performance. The benchmarks not included in the perf-repo are just copied in it, to get the numbers.

> Also, are you explicitly setting JavaScriptEncoder to default in those benchmarks?

The benchmarks in dotnet/performance use the `JavaScriptEncoder` not directly.

For `Test_Escaping` the code is copied into the benchmark-code (as you did in the snippet), based on latest versions available. The labeling is misleading -- my fault, sorry.
In S.T.E.Web it's basically the same code (except the check for `encoder != null`), but I'll update this benchmark and re-run it.

gfoidl commented:
I'll redo the benchmarks, as the code changed a bit with the aim to reduce the overhead (this is somewhat iffy and the results not really stable for small inputs, maybe there's some code-alignment issue).
I think I need a day or two until it's fine...(can't work all the time on this :wink).

gfoidl commented:
Now I got a trace why the numbers for small inputs are so bad, compared to current master.

(Which I couldn't understand so far, as the asm for the sequential path looks quite similar, this PR produces even less instructions therefore, so a difference of 30% felt definitely too much.)

`Test_Escaping`-benchmark-code is
```c#
// setup...not shown

[Benchmark(Baseline = true)]
public unsafe int Master()
{
    fixed (char* ptr = TestStringData.JsonString)
    {
        return DefaultJavaScriptEncoderBasicLatin_master.s_singleton.FindFirstCharacterToEncode(ptr, TestStringData.JsonString.Length);
    }
}

[Benchmark]
public int PR_42073()
{
    fixed (char* ptr = TestStringData.JsonString)
    {
        return DefaultJavaScriptEncoderBasicLatin.s_singleton.FindFirstCharacterToEncode(ptr, TestStringData.JsonString.Length);
    }
}
```
where `DefaultJavaScriptEncoderBasicLatin_master` is the copy from master, so it can be compared in a direct way.

`FindFirstCharacterToEncode` is attributed with `AggressiveInlining` (I'm pretty sure we don't want to put this attribute on this large method anyway).

When I look at the Jit-dumps for a simple program like
```c#
using System;
using System.Text.Encodings.Web;

namespace ConsoleApp1
{
    class Program
    {
        unsafe static int Main()
        {
            string text = "vabvxgyvhoqqfhea¢";

            fixed (char* ptr = text)
            {
                int idx = JavaScriptEncoder.Default.FindFirstCharacterToEncode(ptr, text.Length);
                return idx;
            }
        }
    }
}
```
I can see that `JavaScriptEncoder.Default` gets devirtualized:
```
Querying runtime about current class of field DefaultJavaScriptEncoderBasicLatin.s_singleton (declared as DefaultJavaScriptEncoderBasicLatin)
Field's current class not available

impDevirtualizeCall: Trying to devirtualize virtual call:
    class for 'this' is DefaultJavaScriptEncoderBasicLatin [final] (attrib 21000010)
    base method is TextEncoder::FindFirstCharacterToEncode
    devirt to DefaultJavaScriptEncoderBasicLatin::FindFirstCharacterToEncode -- final class
```
but not inlined
```
[FAILED: target not direct] TextEncoder:FindFirstCharacterToEncode(long,int):int:this
```
(which is ok, as the code to inline is huge).
It's the same for _master_ and _this PR_.

For the `Test_Escaping`-benchmark this means (just showing length := 1)
#### With AggressiveInlining

```
|   Method | TestStringData |     Mean |     Error |    StdDev | Ratio |
|--------- |--------------- |---------:|----------:|----------:|------:|
|   Master |     (1, -1, r) | 3.928 ns | 0.0074 ns | 0.0058 ns |  1.00 |
| PR_42073 |     (1, -1, r) | 4.849 ns | 0.0117 ns | 0.0104 ns |  1.23 |
|          |                |          |           |           |       |
|   Master |   (1, 0, &lt;) | 3.815 ns | 0.0131 ns | 0.0109 ns |  1.00 |
| PR_42073 |   (1, 0, &lt;) | 4.583 ns | 0.0107 ns | 0.0094 ns |  1.20 |
```

#### Without AggressiveInlining

```
|   Method | TestStringData |     Mean |     Error |    StdDev | Ratio |
|--------- |--------------- |---------:|----------:|----------:|------:|
|   Master |     (1, -1, r) | 5.320 ns | 0.0132 ns | 0.0124 ns |  1.00 |
| PR_42073 |     (1, -1, r) | 5.337 ns | 0.0083 ns | 0.0074 ns |  1.00 |
|          |                |          |           |           |       |
|   Master |   (1, 0, &lt;) | 5.277 ns | 0.0119 ns | 0.0105 ns |  1.00 |
| PR_42073 |   (1, 0, &lt;) | 5.093 ns | 0.0081 ns | 0.0068 ns |  0.97 |
```

As said before `FindFirstCharacterToEncode` won't be inlined, so the previous benchmark-results (in the comments above) are not correct, and without `AggressiveInlining` the numbers look quite satifying.
I'll re-run the benchmarks, to get the correct numbers.

---

Aside: when `FindFirstCharacterToEncode` won't be inlined, I don't unterstand why there's such a difference in the benchmarks. BDN calls the benchmark from a delegate in order to prevent inlining-issues (at least is this what I know), so it feels strange here...or the is something really obvious which I don't see right now.

gfoidl commented:
I merged the branch from https://github.com/dotnet/corefx/pull/42064 into this PR (https://github.com/dotnet/corefx/pull/42064#discussion_r338880864), so these changes are also shown right now in the diff.

gfoidl commented on src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/Ssse3Helper.cs:
```diff
@@ -0,0 +1,103 @@
+            return mask;
+        }
+
+        private static readonly Vector128<sbyte> s_nibbleMaskSByte = Vector128.Create((sbyte)0xF);
```
Is there a better name for the mask `0xF`?

ahsonkhan replied:
I have been using the exact unicode character name, but maybe a context/usage specific name would be better:
https://www.fileformat.info/info/unicode/char/000f/index.htm

gfoidl commented on src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/DefaultJavaScriptEncoderBasicLatin.cs:
```diff
@@ -73,122 +73,261 @@ public override bool WillEncode(int unicodeScalar)
 
+            Debug.Assert(textLength >= 0);
+
+
```
Ups, one empty line too much. Will update later...

gfoidl commented on src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/DefaultJavaScriptEncoderBasicLatin.cs:
```diff
@@ -73,122 +73,261 @@ public override bool WillEncode(int unicodeScalar)
+            while (ptr < end);
+
+        AllAllowed:
+            idx = -1;  // All characters are allowed.
```
This comment is somewhat reduntant. I'll remove it, or should it be kept?

gfoidl commented:
Current benchmark-numbers are updated in the top-post.
Now it goes in the right direction 😄

gfoidl commented:
I'll also run benchmarks with SSSE3 disabled, to see if the SSE2 code-path didn't regress.

GrabYourPitchforks commented on src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/DefaultJavaScriptEncoderBasicLatin.cs:
```diff
@@ -73,122 +73,260 @@ public override bool WillEncode(int unicodeScalar)
+            {
+                Debug.Assert(text <= ptr && ptr <= (text + textLength - 2 * Vector128<short>.Count));
+
+                // Load the next 16 characters, combine them to one byte vector
```
Should probably have a comment clarifying what the `packsswb` instruction does: chars that don't cleanly convert to ASCII bytes will get converted to somewhere in the range `[ 0x7F .. 0xFF ]`, which the `NeedsEscaping` method will detect.

gfoidl replied:
Added the comment.

GrabYourPitchforks commented on src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/DefaultJavaScriptEncoderBasicLatin.cs:
```diff
@@ -73,122 +73,260 @@ public override bool WillEncode(int unicodeScalar)
+
+#if NETCOREAPP
+        VectorizedEntry:
+            short* vectorizedEnd = end - 2 * Vector128<short>.Count;
```
Potential underflow. The if check on line 95 validates that we have 8 chars we can read, but this subtraction assumes we're going to read 16 chars at a time.

In reality this shouldn't be a problem because the lowermost addresses of memory should be off-limits, but we should still avoid such underflows in the code.

gfoidl replied:
Good catch.

GrabYourPitchforks commented on src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/Ssse3Helper.cs:
```diff
@@ -0,0 +1,103 @@
+
+            Vector128<sbyte> mask = Sse2.And(bitPositions, bitMask);
+
+            Vector128<sbyte> ne0Gt = Sse2.CompareGreaterThan(mask, s_nullMaskSByte);
```
If you're looking for the equivalent of "not equal to zero", I believe the following will produce more efficient codegen:

```cs
return Sse2.CompareEqual(Sse2.CompareEqual(mask, default), default);
```

(And it'd allow getting rid of the static field.)

gfoidl replied:
Nice! 👍 
It safes the oring of the tmp-comparisons.

So far I didn't know that the JIT treats `default` here the same way as `Vector128<sbyte>.Zero`.

> (And it'd allow getting rid of the static field.)

The static field is intentional, as the JIT hoists it outside the loop.

<details>
  <summary>w/o static ro field</summary>

```asm
; ...
       C4E25900DB           vpshufb  xmm3, xmm4, xmm3
       C4E25100E7           vpshufb  xmm4, xmm5, xmm7
       C5D9DBDB             vpand    xmm3, xmm4, xmm3
       C5D857E4             vxorps   xmm4, xmm4
       C5D057ED             vxorps   xmm5, xmm5
       C5D174DB             vpcmpeqb xmm3, xmm5, xmm3
       C5D974DB             vpcmpeqb xmm3, xmm4, xmm3
       C5F9D7C3             vpmovmskb eax, xmm3
; ...
```
</details>

<details>
  <summary>with static ro field</summary>

```asm
; ...
       C5F9105808           vmovupd  xmm3, xmmword ptr [rax+8]

G_M60692_IG11:      ;; bbWeight=4   ; loop start
; ...
       C4E25100E4           vpshufb  xmm4, xmm5, xmm4
       C4C24900E9           vpshufb  xmm5, xmm6, xmm9
       C5D1DBE4             vpand    xmm4, xmm5, xmm4
       C5B974E4             vpcmpeqb xmm4, xmm8, xmm4
       C5B974E4             vpcmpeqb xmm4, xmm8, xmm4
       C5F9D7C4             vpmovmskb eax, xmm4
; ...
```
</details>

gfoidl commented:
Resolved the merge-conflict with _master_ (src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/Sse2Helper.cs).

GrabYourPitchforks commented:
Test failure is unrelated. I can merge if you have no further changes you wish to make.

gfoidl commented:
I have no further changes to this PR.

GrabYourPitchforks commented:
It's in! Thanks so much for the contribution. :D
